File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.py
class RandomizedSet:

    def __init__(self):
        self.map = {}
        self.nums = []

    def insert(self, val: int) -> bool:
        if val in self.map:
            return False

        self.nums.append(val)
        self.map[val] = len(self.nums) - 1

        return True

    def remove(self, val: int) -> bool:
        if val not in self.map:
            return False

        lastVal = self.nums[-1]
        removeIdx = self.map[val]
        self.nums[removeIdx], self.nums[len(self.nums) - 1] = self.nums[len(self.nums) - 1], self.nums[removeIdx]

        self.map[lastVal] = removeIdx
        del self.map[val]
        self.nums.pop()

        return True
        

    def getRandom(self) -> int:
        rand = random.randint(0, len(self.nums)-1)
        return self.nums[rand]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
class RandomizedSet {
public:
    vector<int> nums;
    unordered_map<int,int> valToIdx;

    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        // check hashmap, return -1 if exists
        if (valToIdx.contains(val)) {
            return false;
        }

        // push back to nums
        // udpate valToIdx
        nums.push_back(val);
        valToIdx[val] = nums.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        // check hashmap, return -1 if exists
        if (!valToIdx.contains(val)) {
            return false;
        }

        // get idx of val
        // get lastVal
        // swap last item with idx
        // valToIdx[lastVal] = idx
        // remove val from hashmap
        // nums pop back
        int idx = valToIdx[val];
        int lastVal = nums.back();
        swap(nums[idx], nums[nums.size() - 1]);

        valToIdx[lastVal] = idx;
        valToIdx.erase(val);
        nums.pop_back();
        return true;
    }
    
    int getRandom() {
        // idx = getRand() % size of num
        // return nums with idx
        int idx = rand() % nums.size();
        return nums[idx];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */